llvm/lib/Target/AIE/AIEHazardRecognizer.cpp

```diff
@@ -243,21 +247,30 @@ void applyFormatOrdering(AIE::MachineBundle &Bundle, const VLIWFormat &Format,
     assert(SlotInfo);
 
     llvm::MachineInstr *Instr = Bundle.at(Slot);
-    if (Instr) {
-      Instr->removeFromBundle();
-      MBB.insert(InsertPoint, Instr);
+    if (!Instr)
+      continue;
+
+    // If the instruction is bundled below a BUNDLE instruction, remove the
+    // latter as well. A new BUNDLE opcode will be created at the end.
+    if (auto BundleStart = getBundleStart(Instr->getIterator());
+        BundleStart->isBundle()) {
+      assert(BundleStart != Instr->getIterator());
+      BundleStart->eraseFromBundle();
```

Comment:
I'm wondering when this occurs, and whether we should not remove  BUNDLEs as a precondition to calling this. Apart from that, I wonder whether there's anybody caring about the format order

Reply:
- This happens when the prologue has VLIW instructions. They get materialized as `BUNDLE`. I could separate that "debundling" in a helper function if you want.
 - The ordering is mostly for stable ref updates at this point. Otherwise slight changes to the `tryCandidate` heuristics can be annoying, because they would change the order of instructions within bundles.

Reply:
Ok. ref order is a good reason. Maybe my main confusion is that this function should live somewhere else.

Reply:
It should. But I blame AIE1 ;)